pyro/infer/trace_elbo.py

```diff
@@ -70,6 +70,46 @@ def loss(self, model, guide, *args, **kwargs):
         warn_if_nan(loss, "loss")
         return loss
 
+    def _differentiable_loss_particle(self, model_trace, guide_trace):
+        elbo_particle = 0
+        surrogate_elbo_particle = 0
+        log_r = None
+
+        # compute elbo and surrogate elbo
+        for name, site in model_trace.nodes.items():
+            if site["type"] == "sample":
+                elbo_particle = elbo_particle + torch_item(site["log_prob_sum"])
+                surrogate_elbo_particle = surrogate_elbo_particle + site["log_prob_sum"]
+
+        for name, site in guide_trace.nodes.items():
+            if site["type"] == "sample":
+                log_prob, score_function_term, entropy_term = site["score_parts"]
+
+                elbo_particle = elbo_particle - torch_item(site["log_prob_sum"])
+
+                if not is_identically_zero(entropy_term):
+                    surrogate_elbo_particle = surrogate_elbo_particle - entropy_term.sum()
+
+                if not is_identically_zero(score_function_term):
+                    if log_r is None:
+                        log_r = _compute_log_r(model_trace, guide_trace)
+                    site = log_r.sum_to(site["cond_indep_stack"])
+                    surrogate_elbo_particle = surrogate_elbo_particle + (site * score_function_term).sum()
+
+        return -elbo_particle, -surrogate_elbo_particle
+
+    def differentiable_loss(self, model, guide, *args, **kwargs):
+        """
+        Computes the surrogate loss that can be differentiated with autograd
+        to produce gradient estimates for the model and guide parameters
+        """
+        surrogate_loss = 0.
+        for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
+            loss_particle, surrogate_loss_particle = self._differentiable_loss_particle(model_trace, guide_trace)
+            surrogate_loss += surrogate_loss_particle / self.num_particles
+        warn_if_nan(surrogate_loss, "loss")
+        return surrogate_loss
```

Comment:
What do you think about wrapping the `(loss, surrogate_loss)` pair in a `torch.autograd.Function` with a `@once_differentiable` `.backward()` method. That would be a little safer for use by multi optimizers, since the 0th and 1st derivatives would be correct and the 2nd derivative would raise an error.

Reply:
Or maybe we should define
```py
def differentiable_loss(...):
    loss, surrogate_loss = ...
    return loss + (surrogate_loss - surrogate_loss.detach())
```

Reply:
I'd forgotten about this since in OED we haven't had any score-function terms yet.  We should probably just allow the DICE estimator in `TraceEnum_ELBO` to be used even when we're not enumerating - I think it currently throws an error in that case.

Reply:
I think you can set `strict_enumeration_warning=False` to use `TraceEnum_ELBO` for non-enumerated models.

Reply:
I'm happy to merge this if you `return loss + (surrogate_loss - surrogate_loss.detach())`. Do you want to make that change or just close this PR and use `TraceEnum_ELBO`?

Reply:
Done